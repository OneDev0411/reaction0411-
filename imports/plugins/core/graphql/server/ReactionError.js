import ExtendableError from "es6-error";

export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
  }
}
