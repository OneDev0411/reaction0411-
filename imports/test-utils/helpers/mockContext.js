const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
