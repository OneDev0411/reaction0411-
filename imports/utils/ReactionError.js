import ExtendableError from "es6-error";

/**
 * @name ReactionError
 * @class
 * @memberof GraphQL
 * @summary A type of error that allows additional details to be provided, which can
 *   then be sent back to the client for a GraphQL request.
 */
export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
    // Newer versions of DDP use this property to signify that an error
    // can be sent back and reconstructed on the calling client.
    this.isClientSafe = true;
    // DDP expects this to be in a property named `reason`
    this.reason = message;
    // DDP expects this to be in a property named `details`
    this.details = eventData;
  }

  clone() {
    return new ReactionError(this.error, this.reason, this.eventData);
  }
}
