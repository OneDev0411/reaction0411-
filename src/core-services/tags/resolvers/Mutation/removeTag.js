import { decodeShopOpaqueId } from "../../../../xforms/shop.js";
import { decodeTagOpaqueId } from "../../../../xforms/tag.js";

/**
 * @name Mutation.removeTag
 * @method
 * @memberof Routes/GraphQL
 * @summary Remove a specified tag
 * @param {Object} parentResult - unused
 * @param {Object} args.input - RemoveTagInput
 * @param {String} args.input.id - id of the tag to remove
 * @param {String} args.input.shopId - shopId of the tag to remove
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @returns {Promise<Object>} RemoveTagPayload
 */
export default async function removeTag(parentResult, { input }, context) {
  const {
    clientMutationId = null,
    id: opaqueTagId,
    shopId: opaqueShopId
  } = input;

  const shopId = decodeShopOpaqueId(opaqueShopId);
  const tagId = decodeTagOpaqueId(opaqueTagId);

  const tag = await context.mutations.removeTag(context, {
    shopId,
    tagId
  });

  return {
    clientMutationId,
    tag
  };
}
