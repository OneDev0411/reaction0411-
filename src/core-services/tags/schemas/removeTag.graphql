"Input for `removeTag` mutation"
input RemoveTagInput {
  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "ID of tag to delete"
  id: ID!

  "The shop that owns the tag"
  shopId: ID!
}

"Response payload for `removeTag` mutation"
type RemoveTagPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The deleted tag"
  tag: Tag!
}

extend type Mutation {
  "Removes an existing tag"
  removeTag(
    "Mutation input"
    input: RemoveTagInput!
  ): RemoveTagPayload!
}
