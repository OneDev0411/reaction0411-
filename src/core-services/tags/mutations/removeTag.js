import ReactionError from "@reactioncommerce/reaction-error";

/**
 * @name Mutation.removeTag
 * @method
 * @memberof Routes/GraphQL
 * @summary Add a tag
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - mutation input
 * @returns {Promise<Object>} RemoveTagPayload
 */
export default async function removeTag(context, input) {
  const { shopId, tagId } = input;
  const { Tags } = context.collections;

  // Check for owner or admin permissions from the user before allowing the mutation
  await context.validatePermissions(`reaction:tags:${tagId}`, "delete", { shopId, legacyRoles: ["owner", "admin"] });

  const tag = await Tags.findOne({ _id: tagId, shopId });
  const { result } = await Tags.deleteOne({ _id: tagId, shopId });

  if (result.n === 0) {
    throw new ReactionError("not-found", "Tag not found");
  }

  return tag;
}
