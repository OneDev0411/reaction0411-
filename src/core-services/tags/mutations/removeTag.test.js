import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import ReactionError from "@reactioncommerce/reaction-error";
import removeTag from "./removeTag.js";

const testShopId = "1234";
const testTagId = "5678";

beforeEach(() => {
  jest.resetAllMocks();
});

test("calls mutations.removeTag and returns the RemoveTagPayload on success", async () => {
  mockContext.checkPermissions.mockReturnValueOnce(Promise.resolve(null));
  mockContext.collections.Tags.deleteOne.mockReturnValueOnce({ result: { ok: 1 } });
  mockContext.collections.Tags.findOne.mockReturnValueOnce({
    shopId: testShopId,
    tagId: testTagId
  });

  const input = {
    input: {
      shopId: testShopId,
      tagId: testTagId
    }
  };
  const result = await removeTag(mockContext, input);

  expect(result).toEqual({
    shopId: testShopId,
    tagId: testTagId
  });
  expect(mockContext.collections.Tags.deleteOne).toHaveBeenCalled();
});

test("calls mutations.removeTag and throws for non admins", async () => {
  mockContext.checkPermissions.mockImplementation(() => {
    throw new ReactionError("access-denied", "Access Denied");
  });
  mockContext.collections.Tags.deleteOne.mockReturnValueOnce({ result: { ok: 1 } });

  const result = removeTag(mockContext, {});
  expect(result).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.collections.Tags.deleteOne).not.toHaveBeenCalled();
});
