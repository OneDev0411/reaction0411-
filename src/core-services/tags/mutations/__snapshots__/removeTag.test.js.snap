// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls mutations.removeTag and throws for non admins 1`] = `"Access Denied"`;
